### ISSUE TITLE
Ignoring coverage using range of lines doesn't always work

### ISSUE THREAD

**Issue by abhaysood** [flutter/flutter#83425]
# Issue 1
`coverage:ignore-start` and  `coverage:ignore-end` do not ignore content of lambda functions.
For the sample app below, I've ignored the `_incrementCounter` method.  This is the result:

![image](https://user-images.githubusercontent.com/3646627/119702038-a5ac1700-be72-11eb-9a57-cf51757ee867.png)

# Issue 2
After trying out on multiple cases I found that `coverage:ignore-start` and  `coverage:ignore-end` does not work when the file contains multiple classes. For example in the following example the `debugFillProperties` function gets ignored only if the widget is present in a separate file.

![image](https://user-images.githubusercontent.com/3646627/119702722-66ca9100-be73-11eb-917f-f308ae734c39.png)

### Command to generate test coverage
```
flutter test --coverage
```

# Sample App
```
void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();

  // coverage:ignore-start
  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(IntProperty("name", 1));
  }
// coverage:ignore-end
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  // coverage:ignore-start
  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  // coverage:ignore-end

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}
```

# Test
```
void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
  });
}
```

# Flutter doctor
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel unknown, 2.2.0, on Mac OS X 10.15.4 19E287 darwin-x64, locale en-IN)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.1)
[✓] Android Studio (version 4.2)
[✓] IntelliJ IDEA Ultimate Edition (version 2021.1.1)
[✓] VS Code (version 1.53.0)
[✓] Connected device (1 available)

• No issues found!
```

**Comment by iapicca** [comment]
> ### Issue 2
> ![image](https://user-images.githubusercontent.com/3646627/119702722-66ca9100-be73-11eb-917f-f308ae734c39.png)

it's an absolute mystery to me why `super` counts as "line to cover", what am I supposed to test exactly?

**Comment by TahaTesser** [comment]
Hi @abhaysood 
Can you please provide the exact command you used to run the test?
Thank you

**Comment by iapicca** [comment]
@TahaTesser 
I think this [setup](https://github.com/flutter/flutter/issues/76544#issuecomment-784013652) can help

**Comment by abhaysood** [comment]
@TahaTesser I used `flutter test --coverage`. Updated the original issue with this detail.

**Comment by TahaTesser** [comment]
@iapicca 
Thanks! 

@abhaysood 
I was able to create a coverage report based on https://github.com/flutter/flutter/issues/83425#issuecomment-849522371
<img width="893" alt="Screenshot 2021-05-27 at 2 09 45 PM" src="https://user-images.githubusercontent.com/48603081/119816293-34fa0e80-bef5-11eb-9492-af6a3103c62b.png">



 however `coverage:ignore-start and coverage:ignore-end do not ignore content of lambda functions.`hiow are you ignoring with just `flutter test --coverage`?

**Comment by TahaTesser** [comment]
## Issue 

### Updated code sample


<details> 
<summary>minimal code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  // coverage:ignore-start
  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }
  // coverage:ignore-end

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}

``` 

</details>

### Issue 2

<details> 
<summary>minimal code sample</summary> 

```dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

// import 'second.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Material App',
      theme: ThemeData.dark(),
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Material App Bar'),
      ),
      floatingActionButton: FloatingActionButton(
        key: Key('goToSecond'),
        child: Icon(Icons.add),
        onPressed: () => Navigator.push(
            context, MaterialPageRoute(builder: (_) => Second())),
      ),
    );
  }
}

class Second extends StatefulWidget {
  @override
  State<Second> createState() => _SecondState();
}

class _SecondState extends State<Second> with Diagnosticable {
  // coverage:ignore-start
  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(IntProperty('name', 1));
  }
  // coverage:ignore-end

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Second screen'),
      ),
    );
  }
}

``` 

</details>

```yaml
dev_dependencies:
  flutter_test:
    sdk: flutter

  # The "flutter_lints" package below contains a set of recommended lints to
  # encourage good coding practices. The lint set provided by the package is
  # activated in the `analysis_options.yaml` file located at the root of your
  # package. See that file for information about deactivating specific lint
  # rules and activating additional ones.
  flutter_lints: ^1.0.0
  test_coverage: ^0.5.0
```

### prerequisite
`brew install lcov` (macOS) 

### Command
```
flutter test --coverage
format_coverage --lcov --check-ignore  --in=coverage --out=coverage.lcov --packages=.packages --report-on=lib
genhtml -o coverage/html coverage/lcov.info
```
Open `app/coverage/html/index.html` to view results


Result:

**Issue 1**
I can see _increment method is not getting ignored

**Issue 2**
Doesn't reproduce in single or multi-class

<details> 
<summary>flutter doctor -v</summary> 
	
```console
[✓] Flutter (Channel master, 2.3.0-13.0.pre.187, on macOS 11.3.1 20E241 darwin-x64, locale en-GB)
    • Flutter version 2.3.0-13.0.pre.187 at /Users/tahatesser/Code/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision a96e763155 (8 hours ago), 2021-05-26 23:14:02 -0400
    • Engine revision a832502eaf
    • Dart version 2.14.0 (build 2.14.0-153.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.5, Build version 12E262
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2021.1.1)
    • IntelliJ at /Volumes/Extreme/IntelliJ IDEA CE.app
    • Flutter plugin version 56.0.5
    • Dart plugin version 211.7233

[✓] VS Code (version 1.56.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.22.0

[✓] Connected device (3 available)
    • iPhone 12 (mobile) • D72A5DAA-24E6-421B-B4F5-7CFAFB498040 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • macOS (desktop)    • macos                                • darwin-x64     • macOS 11.3.1 20E241
      darwin-x64
    • Chrome (web)       • chrome                               • web-javascript • Google Chrome 90.0.4430.212

• No issues found!

```

</details>

**Comment by abhaysood** [comment]
Thanks a lot Taha!

Is using [test_coverage](https://pub.dev/packages/test_coverage) package recommended? It seems like the only difference between my setup and yours.

**Comment by TahaTesser** [comment]
Hi @abhaysood 
It's just to used to generate a report, it shouldn't affect the outcome;. 

Can you please code for Issue 2?
Thanks

**Comment by abhaysood** [comment]
Try this for issue 2. I think the issue is not what I thought originally. It's not just having multiple classes in the file. It's more nuanced. In the following sample.

# Result
![image](https://user-images.githubusercontent.com/3646627/119945316-e6ab4500-bfb2-11eb-953d-6e76270f2fe7.png)


# Sample app

```dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Material App',
      theme: ThemeData.dark(),
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Material App Bar'),
      ),
      floatingActionButton: FloatingActionButton(
        key: Key('goToSecond'),
        child: Icon(Icons.add),
        onPressed: () => Navigator.push(
            context, MaterialPageRoute(builder: (_) => Second())),
      ),
    );
  }

  // coverage:ignore-start
  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(IntProperty('name', 1));
  }
// coverage:ignore-end
}

class Second extends StatefulWidget {
  @override
  State<Second> createState() => _SecondState();
}

class _SecondState extends State<Second> with Diagnosticable {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Second screen'),
      ),
    );
  }

  // coverage:ignore-start
  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(IntProperty('name', 1));
  }
// coverage:ignore-end
}

class MyThemeData with Diagnosticable {
  // coverage:ignore-start
  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(IntProperty('name', 1));
  }
// coverage:ignore-end
}
```


# Test
```dart
void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('Material App Bar'), findsOneWidget);
  });
}
```

**Comment by TahaTesser** [comment]
@abhaysood 
Thanks for the code sample, I can replicate it too

<details> 
<summary>flutter doctor -v</summary> 
	
```console
[✓] Flutter (Channel master, 2.3.0-17.0.pre.111, on macOS 11.3.1 20E241
    darwin-x64, locale en-GB)
    • Flutter version 2.3.0-17.0.pre.111 at
      /Users/tahatesser/Code/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 6a38994cb8 (7 hours ago), 2021-05-28 02:09:02 -0400
    • Engine revision ecfd7bdffe
    • Dart version 2.14.0 (build 2.14.0-162.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.5, Build version 12E262
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2021.1.1)
    • IntelliJ at /Volumes/Extreme/IntelliJ IDEA CE.app
    • Flutter plugin version 56.0.5
    • Dart plugin version 211.7233

[✓] VS Code (version 1.56.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.22.0

[✓] Connected device (4 available)
    • Taha’s iPhone (mobile) • 00008020-001059882212002E            • ios
      • iOS 14.6
    • iPhone 12 (mobile)     • D72A5DAA-24E6-421B-B4F5-7CFAFB498040 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • macOS (desktop)        • macos                                • darwin-x64
      • macOS 11.3.1 20E241 darwin-x64
    • Chrome (web)           • chrome                               •
      web-javascript • Google Chrome 90.0.4430.212

• No issues found!

```

</details>

**Comment by liamappelbe** [comment]
This is caused by a bug that has been fixed: https://github.com/dart-lang/coverage/pull/347

The fix is in coverage version 1.1.0, which isn't published yet.
